docs/lsp.md

```diff
@@ -0,0 +1,480 @@
+
+At the moment this is possible via the experimental `DocumentSource.Custom` parameter. But hopefully we will be able to use a new `FSharpProjectSnapshot` model (from [#15179](https://github.com/dotnet/fsharp/pull/15179)) which combines all the necessary inputs into a single object.
+
+
+## LSP library
+
+The library needs to support all the features we want to implement including the ability to add custom F# ones. It also shouldn't incur an unnecessary performance overhead.
+
+### Candidates
+
+| Library  | Pros | Cons |
+| ---  | --- | --- |
+| We write our own | + Complete control over everything | - A lot of extra work |
+| [Ionide LanguageServerProtocol](https://github.com/ionide/LanguageServerProtocol) | + Battle tested with Ionide and FSAC + Natural F# API | - Might not be able to fully customize it to our need |
+| [Roslyn CLaSP](https://github.com/dotnet/roslyn/blob/main/src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/README.md) | + Should be performant + Future support | - Not stable API - Unnatural C# API |
+| [OmniSharp C# LSP](https://github.com/OmniSharp/csharp-language-server-protocol)  |  | - Unnatural C# API |
+
+It probably makes sense to start with an existing library which should be able to get us to a working product and only then consider writing our own in case it's the only way to enable more functionality or performance.
+
+## LSP client (?)
+
+- Is there any code that can be shared by various clients?
+- Most likely not, since each client will be specific to its platform and project system
+
+
+## Diagnostic mode
+
+We will be offering the **Pull Diagnostics** mode where the client will request diagnostics for a particular document. This should be easier to implement and probably more practical anyway.
+
+We could experiment with **Publish Diagnostics** in combination with "keep checking the whole solution in background" option, if we decide to implement it on the server side. But this is a low priority item.
+
+
+## LSIF
+
+We should eventually add support for LSIF. At the moment it looks like we would have to construct it manually (as in, no helpful libraries are available).
+
+This is probably lower priority than anything else.
+
+It might also be a good opportunity to re-think how we store symbols for purposes of finding references, go to definition, or code search (go to symbol in workspace by name/query).
+
+
+## Performance
+
+Areas to pay attention to with regards to performance
+
+- Serialization. Hopefully this will be handled reasonably by a library.
+- Spamming too many requests. Though this will be mainly up to the client, we could think about throttling or debouncing.
+- Caching. This will be mostly up to FCS but we can send hints about what can be cleaned up.
+
+
+## Analyzers & Refactoring support
+
+These will hopefully not require any special treatment and can just use existing LSP APIs.
```

Comment:
One thing that the C# Dev Kit team has found is that there's not a lot of protocol-level support for performing the 'same kind' of fix at a whole-document or whole-project level, so if you want that you may have some custom work.